### ISSUE TITLE
AvgPool2D output shapes are inconsistent when ceil_mode=True

### ISSUE THREAD

**Issue by aws-stdun** [pytorch/pytorch#88144]
### 🐛 Describe the bug

Certain inputs produce unexpected output shapes from [AvgPool2D](https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html) when `ceil_mode=True`.

## Example 1

```python
>>> t = torch.ones((1, 3, 32, 32))
>>> pool = torch.nn.AvgPool2d(8, stride=4, ceil_mode=True)
>>> pool(t).shape
torch.Size([1, 3, 7, 7])
```

This is surprising behavior because there are 8 possible windows with `stride=4` (not 7) for both height and width, each starting from the following indexes:
```
0
4
8
12
16
20
24
28
```

The final window at index 28 is a partial window, but should be included when `ceil_mode=True`, otherwise `ceil_mode` seems to serve no purpose. (This window does not start in a padding region as there is no padding used here.)

## Example 2
```
>>> t = torch.ones((1, 3, 4, 4))
>>> pool = torch.nn.AvgPool2d(3, stride=3, ceil_mode=True)
>>> pool(t).shape
torch.Size([1, 3, 2, 2])
```

With `stride=3`, we have 2 potential start positions:
```
0
3
```

Torch chooses to use both of them. But based on the behavior observed in Example 1 (where the last frame was discarded) we would have expected output shape of `[1, 3, 1, 1]`.

### Versions

PyTorch version: 1.12.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Amazon Linux 2 (x86_64)
GCC version: (GCC) 7.3.1 20180712 (Red Hat 7.3.1-15)
Clang version: Could not collect
CMake version: version 3.18.4
Libc version: glibc-2.2.5

Python version: 3.7.10 (default, Jun  3 2021, 00:02:01)  [GCC 7.3.1 20180712 (Red Hat 7.3.1-13)] (64-bit runtime)
Python platform: Linux-5.4.209-129.367.amzn2int.x86_64-x86_64-with-glibc2.2.5
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.20.0
[pip3] torch==1.12.1
[pip3] torch-neuron==1.12.1.2.0.0.0
[conda] Could not collect

**Comment by samdow** [comment]
Adding high pri for silent correctness (shape inconsistency). I can repro this on master and default is no padding so that shouldn't impact it

**Comment by jgong5** [comment]
@Guobing-Chen

**Comment by Guobing-Chen** [comment]
I will take a look at this issue

**Comment by cpuhrsch** [comment]
Good next steps here could be (as discussed in triage review).
1. Update error message on CPU to match CUDA (which is more informative).
2. Update documentation to describe the conditions under which this issue is thrown. For example, remove or update the existing note as appropriate.
3. Revisit functionality and whether this behavior is reasonable.

**Comment by Guobing-Chen** [comment]
> Good next steps here could be (as discussed in triage review).
> 
> 1. Update error message on CPU to match CUDA (which is more informative).
> 2. Update documentation to describe the conditions under which this issue is thrown. For example, remove or update the existing note as appropriate.
> 3. Revisit functionality and whether this behavior is reasonable.

Good suggestions. But the 1st one may not quite apply as the wrong output shape issue does not always trigger error report or crash or so. As the reported two cases here, we just get not expected shapes but everything still moves on. I am working on the fix as in #88421 but seems it requires much more time as we are correcting output shapes which breaks several other places' assumption of pool/unpool shape calculation. Should we follow the 2nd suggestion directly to note this issue in doc firstly in case we need more time to work on a real fix? If yes, can someone point me the right place of doc that suitable for this?

**Comment by Valentine233** [comment]
I could successfully reproduce the results.
```
Example 1
>>> t = torch.ones((1, 3, 32, 32))
>>> pool = torch.nn.AvgPool2d(8, stride=4, ceil_mode=True)
>>> pool(t).shape
torch.Size([1, 3, 7, 7])
>>> pool = torch.nn.AvgPool2d(8, stride=4, ceil_mode=False)
>>> pool(t).shape
torch.Size([1, 3, 7, 7])

Example 2
>>> t = torch.ones((1, 3, 4, 4))
>>> pool = torch.nn.AvgPool2d(3, stride=3, ceil_mode=True)
>>> pool(t).shape
torch.Size([1, 3, 2, 2])
>>> pool = torch.nn.AvgPool2d(3, stride=3, ceil_mode=False)
>>> pool(t).shape
torch.Size([1, 3, 1, 1])
```

From my point of view, current results make sense and do not have any problem. I suppose that ceil_mode is introduced to handle the case where the last several indices of input could not be calculated for pooling. It adds paddings at the end to make sure every input index be used instead of being dropped.

For Example 1, the 8 window candidates are
```
1: 0-7
2: 4-11
3: 8-15
4: 12-19
5: 16-23
6: 20-27
7: 24-31
8: 28-35
```
We notice that the ending index of the 7th window is 31, which is exactly the last index of input. In this case, every input index is already used with the first 7 windows. Hence, the 8th window seems redundant.

Let's see now the definition of [AvgPool2d](https://pytorch.org/docs/stable/generated/torch.nn.AvgPool2d.html). The parameter ceil_mode (bool) – when True, will use ceil instead of floor to compute the output shape. The formula of output shape is as follows.
<img width="500" alt="image" src="https://user-images.githubusercontent.com/23010269/230261292-8b0344e9-1ee0-469b-b97c-7224c33139b0.png">

We return back to the Example 1, H/W of output is exactly 7 before doing floor or ceil. Thus, according to the formula, the output would keep the shape of [1, 3, 7, 7] whether ceil_mode is on or not.
